org.metafacture.fix/src/main/java/org/metafacture/metafix/RecordTransformer.java

```diff
@@ -0,0 +1,187 @@
+import org.metafacture.metafix.FixPredicate.Quantifier;
+import org.metafacture.metafix.fix.Do;
+import org.metafacture.metafix.fix.ElsIf;
+import org.metafacture.metafix.fix.Else;
+import org.metafacture.metafix.fix.Expression;
+import org.metafacture.metafix.fix.Fix;
+import org.metafacture.metafix.fix.If;
+import org.metafacture.metafix.fix.MethodCall;
+import org.metafacture.metafix.fix.Options;
+import org.metafacture.metafix.fix.Unless;
+
+import com.google.common.collect.ImmutableMap;
+import org.eclipse.emf.common.util.EList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Transform a record using a Fix
+ *
+ * @author Fabian Steeg
+ *
+ */
+class RecordTransformer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RecordTransformer.class);
+
+    private Fix fix;
+    private Map<String, Object> record;
+    private Map<String, String> vars;
+
+    RecordTransformer(final Map<String, Object> record, final Map<String, String> vars, final Fix fix) {
+        this.record = new LinkedHashMap<>(record);
+        this.vars = vars;
+        this.fix = fix;
+    }
+
+    Map<String, Object> transform() {
+        processSubexpressions(fix.getElements());
+        return record;
+    }
+
+    Map<String, Object> getRecord() {
+        return record;
+    }
+
+    private String resolveVars(final String string) {
+        return string == null ? null : StringUtil.format(string, Metafix.VAR_START, Metafix.VAR_END, false, vars);
+    }
+
+    private void processSubexpressions(final List<Expression> expressions) {
+        for (final Expression sub : expressions) {
+            final EList<String> params = sub.getParams();
+            if (sub instanceof Do) {
+                processBind((Do) sub, params);
+            }
+            else if (sub instanceof If) {
+                processIf((If) sub, params);
+            }
+            else if (sub instanceof Unless) {
+                processUnless((Unless) sub, params);
+            }
+            else {
+                processFunction(sub, params);
+            }
+        }
+    }
+
+    private void processBind(final Do theDo, final EList<String> params) {
+        if (theDo.getName().equals("list")) { // TODO impl multiple binds via FixBind enum
+            final Map<String, String> options = options(theDo.getOptions());
+            final Map<String, Object> fullRecord = new LinkedHashMap<>(record);
+            final Object values = FixMethod.find(record, FixMethod.split(options.get("path")));
+            Metafix.asList(values).stream().filter(val -> val != null).forEach(val -> {
+                // for each val, bind the current record/scope/context to the given var name:
+                record = new LinkedHashMap<>(ImmutableMap.of(options.get("var"), val));
```

Comment:
This prevents access to the actual record. Is this intentional?

Reply:
My idea was that in the scope of the bind, the record (as the subexpressions see it) should only contain the bound value. I'm not sure why though, that might very well make no sense.

Reply:
I wasn't able to find it documented anywhere, but Catmandu seems to preserve access to the whole record:

```perl
$fixes = <<EOF;
do list(path:foo,var:loop)
  copy_field(test,loop.baz)
end
EOF

$fixer = Catmandu::Fix->new(fixes => [$fixes]);

is_deeply $fixer->fix({foo => [{bar => 1}, {bar => 2}], test => 42}),
    {foo => [{bar => 1, baz => 42}, {bar => 2, baz => 42}], test => 42},
    'binding scope testing';
```

Reply:
I agree, the full record should be available. I've opened https://github.com/metafacture/metafacture-fix/issues/66 since this does require some work that I probably won't have time to do this week (simply using the full record with the current implementation results in overwriting previous values under the same key, which is covered by a test).